lib/ArduinoNative/SocketServer.cpp

```diff
@@ -0,0 +1,263 @@
+
+/******************************************************************************
+ * Local Variables
+ *****************************************************************************/
+
+/******************************************************************************
+ * Public Methods
+ *****************************************************************************/
+
+SocketServer::SocketServer() : m_clientSocket(INVALID_SOCKET), m_listenSocket(INVALID_SOCKET)
+{
+}
+
+SocketServer::~SocketServer()
+{
+    close();
+}
+
+bool SocketServer::init(uint16_t port, uint8_t maxConnections)
+{
+    bool             success = true;
+    int              result;
+    struct addrinfo  hints;
+    struct addrinfo* addrInfo = nullptr;
+
+    memset(&hints, 0, sizeof(struct addrinfo));
+
+    hints.ai_family   = AF_INET;
+    hints.ai_socktype = SOCK_STREAM;
+    hints.ai_protocol = IPPROTO_TCP;
+    hints.ai_flags    = AI_PASSIVE;
+
+#ifdef _WIN32
+    WSADATA wsaData;
+
+    /* Initialize Winsock */
+    result = WSAStartup(MAKEWORD(2, 2), &wsaData);
+    if (0 != result)
+    {
+        printf("WSAStartup failed with error: %d\n", result);
+        return false;
+    }
+
+#endif
+
+    /* Resolve the server address and port */
+    result = getaddrinfo(nullptr, std::to_string(port).c_str(), &hints, &addrInfo);
+    if (0 != result)
+    {
+        printf("getaddrinfo failed with error: %d\n", result);
+        close();
+        return false;
+    }
+
+    /* Create a SOCKET for the server to listen for client connections. */
+    m_listenSocket = socket(addrInfo->ai_family, addrInfo->ai_socktype, addrInfo->ai_protocol);
+    if (INVALID_SOCKET == m_listenSocket)
+    {
+        printf("socket failed\n");
+        freeaddrinfo(addrInfo);
+        close();
+        return false;
+    }
+
+    /* Setup the TCP listening socket */
+    result = bind(m_listenSocket, addrInfo->ai_addr, static_cast<int>(addrInfo->ai_addrlen));
+    if (SOCKET_ERROR == result)
+    {
+        printf("bind failed\n");
+        freeaddrinfo(addrInfo);
+        close();
+        return false;
+    }
+
+    freeaddrinfo(addrInfo);
+
+    result = listen(m_listenSocket, maxConnections);
+    if (SOCKET_ERROR == result)
+    {
+        printf("listen failed\n");
+        close();
+        return false;
+    }
+
+    return true;
+}
+
+size_t SocketServer::sendMessage(const uint8_t* buf, uint16_t length)
+{
+    size_t bytesSent = 0;
+
+    /* Echo the buffer back to the sender */
+    if (INVALID_SOCKET != m_clientSocket)
+    {
+        int result = send(m_clientSocket, reinterpret_cast<const char*>(buf), length, 0);
+        if (SOCKET_ERROR == result)
+        {
+            printf("send failed\n");
+            /* Error on the socket. Client is now invalid. */
+            m_clientSocket = INVALID_SOCKET;
+        }
+        else
+        {
+            bytesSent = result;
+        }
+    }
+
+    return bytesSent;
+}
+
+uint32_t SocketServer::available()
+{
+    this->processRx();

```

Comment:
Is the call to processRx() really necessary here? Because processRx() shall be called periodically anyway.

Reply:
It is not strictly necessary, but I found it was the best place to have it. processRx() shall be called periodically, but only the Serial_ Class has an instance of the SocketServer, the there is no method called periodically inside Serial_.
What would you propose as an alternative?